src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -61,30 +63,79 @@ internal sealed class State
                 SingleReader = true,
                 SingleWriter = true,
             });
+
+            public void RemoveStream(MsQuicStream stream)
+            {
+                bool releaseHandles;
+                lock (this)
+                {
+                    StreamCount--;
+                    releaseHandles = _closing && StreamCount == 0;
+                }
+
+                if (releaseHandles)
+                {
+                    Handle?.Dispose();
+                    if (StateGCHandle.IsAllocated) StateGCHandle.Free();
+                }
+            }
+
+            public bool TryQueueNewStream(SafeMsQuicStreamHandle streamHandle, QUIC_STREAM_OPEN_FLAGS flags)
+            {
+                var stream = new MsQuicStream(this, streamHandle, flags);
+                if (AcceptQueue.Writer.TryWrite(stream))
+                {
+                    return true;
+                }
+                else
+                {
+                    stream.Dispose();
+                    return false;
+                }
+            }
+
+            public bool TryAddStream(MsQuicStream stream)
+            {
+                lock (this)
+                {
+                    if (_closing)
+                    {
+                        return false;
+                    }
+
+                    StreamCount++;
+                    return true;
+                }
+            }
+
+            // This is called under lock from connection dispose
+            public void SetClosing()
+            {
+                _closing = true;
```

Comment:
Without doing this, platform is technically allowed to never (not even eventually) make the change visible to other CPUs. I guess we'd observe this problem on ARM.

```suggestion
                lock(this)
                {
                    _closing = true;
                }
```

Reply:
updated.